src/main/java/io/smartcat/ranger/core/ValueProxy.java

```diff
@@ -0,0 +1,59 @@
+package io.smartcat.ranger.core;
+
+/**
+ * Proxy around value that can cache value and can reset cache.
+ *
+ * @param <T> Type this value would evaluate to.
+ */
+public class ValueProxy<T> implements Value<T> {
+
+    private Value<T> delegate;
+    private T evaluatedValue;
+    private boolean isEvaluated;
+
+    /**
+     * Constructs proxy without delegate.
+     */
+    public ValueProxy() {
```

Comment:
Can we force creating of ValueProxy with delegate always (loose this constructor). That will make this class a lot simpler and I think it does not make sense to have proxy without delegate.

We can then have only constructor which accepts delegate, loose null check and loose setter.